R/benchmarks.R

```diff
@@ -2,23 +2,52 @@
 #'
 #' @description Tests that the shapley algorithm converges.
 #' @param row.nr Index for the observation of interest.
-#' @param iterations Amount of iterations.
-#' @param shapley.iterations Amount of the iterations within the shapley
-#'   function.
-#' @return shapley value as a data.frame with col.names and their corresponding
-#'   effects.
-test.covergence = function(row.nr, iterations = 20, shapley.iterations = 1) {
-  values = rep(0, times = iterations)
-  for(i in 1:iterations) {
-    values[i] = sum(shapley(row.nr, iterations = shapley.iterations))
-  }
+#' @param convergence.iterations Amount of calls of the shapley function.
+#' @param iterations Amount of the iterations within the shapley function.
+#' @return shapley values as a data.frame or a plot
+#'
+test.convergence = function(row.nr=2, convergence.iterations = 20, iterations = 20, task = iris.task,
+                            model = train(makeLearner("classif.lda", predict.type = "prob"), iris.task),
+                            return.value = "values") {
 
-  return(values)
-}
+  assert_number(iterations, convergence.iterations, row.nr)
+  assert_choice(return.value, c("plot", "values"))
+  assert_class(model, "WrappedModel")
+
+  data = getTaskData(task)[row.nr,]
+  prediction = getPredictionData(data, model, task)
+  prediction.class = names(prediction[BBmisc::getMaxIndex(prediction)])
 
-#a = test.covergence(196, iterations = 20)
-#a.cum = cumsum(a) / seq_along(a)
-#a.sub = a
-#ggplot() +
-#  geom_point(aes(x = seq_along(a), y = a)) +
-#  geom_line(aes(x = seq_along(a), y = cumsum(a)/seq_along(a)), color = "red")
+  values = rep(0, times = convergence.iterations)
+  for(i in 1:convergence.iterations) {
+    if(getTaskType(task) == "classif"){
+      shap = shapley(row.nr, task, model, iterations)
+      values[i] = shap$"_Class"[BBmisc::getMaxIndex(rowSums(shap[,getTaskFeatureNames(task)]))]
+    }
+    else if(getTaskType(task) == "regr"){
+      shap = shapley(row.nr, task, model, iterations)
+      truth = getPredictionTruth(prediction)
+      values[i] = sum(shap[,getTaskFeatureNames(task)]) + truth
+    }
+  }
+  if(return.value == "plot") {
+    plot = ggplot() +
+      geom_point(aes(x = seq_along(values), y = values, colour = "Sum of Shapley values")) +
+      geom_line(aes(x = seq_along(values), y = cumsum(values)/seq_along(values), colour = "Moving Averages")) +
+      geom_line(aes(x = seq(1:convergence.iterations), y = rep(response, convergence.iterations), colour = "Prediction")) +
+      scale_colour_discrete(name = NULL) + labs(x = "Convergence iterations", y = "Shapley value") +
+      theme(legend.position="bottom")
+    return(plot)
+  }
+  if(return.value == "values") {
+    if(getTaskType(task) == "classif"){
+      result = plyr::count(as.data.frame(values))

```

Comment:
Can you solve it without an extra package? According to our previous lessons we should avoid dependiencies as far as it is possible. The following (not complete) code shows, that we can get fine with the basic methods:
`> a = as.data.frame(rbind("A", "A", "B", "D", "A", "D"))
> names(a) = "values"
> summary(a$values)
A B D 
3 1 2 `
And with `summary(a$values)["A"]` you can get `A 3` as a result. So I think this should be the same as plyr::count but needs a little bit more code

Reply:
Done.